src/test/java/com/google/api/generator/gapic/composer/ResourceNameHelperClassComposerTest.java

```diff
@@ -0,0 +1,170 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.api.generator.gapic.composer;
+
+import static com.google.common.truth.Truth.assertThat;
+import static junit.framework.Assert.assertEquals;
+
+import com.google.api.generator.engine.writer.JavaWriterVisitor;
+import com.google.api.generator.gapic.model.GapicClass;
+import com.google.api.generator.gapic.model.Message;
+import com.google.api.generator.gapic.model.ResourceName;
+import com.google.api.generator.gapic.model.Service;
+import com.google.api.generator.gapic.protoparser.Parser;
+import com.google.protobuf.Descriptors.FileDescriptor;
+import com.google.protobuf.Descriptors.ServiceDescriptor;
+import com.google.showcase.v1beta1.EchoOuterClass;
+import java.util.Arrays;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+import org.junit.Before;
+import org.junit.Test;
+
+public class ResourceNameHelperClassComposerTest {
+  private ServiceDescriptor echoService;
+  private FileDescriptor echoFileDescriptor;
+
+  @Before
+  public void setUp() {
+    echoFileDescriptor = EchoOuterClass.getDescriptor();
+    echoService = echoFileDescriptor.getServices().get(0);
+    assertEquals(echoService.getName(), "Echo");
+  }
+
+  @Test
+  public void parseTokenHierarchy_basic() {
+    List<String> patterns =
+        Arrays.asList(
+            "projects/{project}/locations/{location}/autoscalingPolicies/{autoscaling_policy}",
+            "projects/{project}/regions/{region}/autoscalingPolicies/{autoscaling_policy}",
+            "projects/{project}/autoscalingPolicies/{autoscaling_policy}");
+    List<List<String>> tokenHierarchies =
+        ResourceNameHelperClassComposer.parseTokenHierarchy(patterns);
+    assertEquals(3, tokenHierarchies.size());
+    assertThat(tokenHierarchies.get(0))
+        .containsExactly("project", "location", "autoscaling_policy");
+    assertThat(tokenHierarchies.get(1)).containsExactly("project", "region", "autoscaling_policy");
+    assertThat(tokenHierarchies.get(2)).containsExactly("project", "autoscaling_policy");
+  }
+
+  @Test
+  public void parseTokenHierarchy_singletonCollection() {
+    List<String> patterns =
+        Arrays.asList(
+            "projects/{project}/agent/sessions/{session}",
+            "projects/{project}/agent/environments/{environment}/users/{user}/sessions/{session}");
+    List<List<String>> tokenHierarchies =
+        ResourceNameHelperClassComposer.parseTokenHierarchy(patterns);
+    assertEquals(2, tokenHierarchies.size());
+    assertThat(tokenHierarchies.get(0)).containsExactly("project", "session");
+    assertThat(tokenHierarchies.get(1))
+        .containsExactly("project", "environment", "user", "session");
+  }
+
+  @Test
+  public void parseTokenHierarchy_invalidPatterns() {
+    List<String> patterns =
+        Arrays.asList(
```

Comment:
For confirmation, we do not want throw errors even if this pattern is invalid, right?

Also how about the pattern with wildcard in token like `project/{project=**}`? Or this is not in our consideration yet. Thanks

Reply:
No, we don't want to throw errors here. The wildcard patterns will be handled by L237 in the corresponding class.